subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/BeanNode.java

```diff
@@ -19,32 +19,132 @@
+        public Iterator<BeanNode> getIterator() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class IterableBeanNode extends BeanNode {
+        private final Iterable<?> iterable;
+
+        public IterableBeanNode(@Nullable String propertyName, Iterable<?> iterable) {
+            super(propertyName, iterable);
+            this.iterable = iterable;
+        }
+
+        @Override
+        public boolean isIterable() {
+            return true;
+        }
+
+        @Override
+        public Iterator<BeanNode> getIterator() {
+            return Iterators.transform(iterable.iterator(), new Function<Object, BeanNode>() {
+                private int count = 0;
+                private Set<String> seenNames = new HashSet<String>();
+
+                @Override
+                public BeanNode apply(@Nullable Object input) {
+                    String propertyName = determinePropertyName(input);
+                    Preconditions.checkState(seenNames.add(propertyName),
+                        "Nested iterables can only contain beans with unique names. Duplicate name: '%s'.",
+                        getQualifiedPropertyName(propertyName));
+                    return createChildNode(propertyName, input);
+                }
+
+                private String determinePropertyName(@Nullable Object input) {
+                    if (input instanceof Named) {
+                        return ((Named) input).getName();
+                    }
+                    return "$" + count++;
+                }
+            });
+        }
+
+        @Override
+        public Object getBean() {
+            return iterable;
+        }
     }
 
-    @Override
-    public Iterator<BeanNode> getIterator() {
-        return Iterators.transform(((Iterable<?>) bean).iterator(), new Function<Object, BeanNode>() {
-            private int count = 0;
-
-            @Override
-            public BeanNode apply(@Nullable Object input) {
-                String childPropertyName = getQualifiedPropertyName("$" + count++);
-                return new BeanNode(childPropertyName, Preconditions.checkNotNull(input, "Null is not allowed as nested property '" + childPropertyName + "'"));
-            }
-        });
+    private static class MapBeanNode extends BeanNode {
+        private final Map<?, ?> map;
+
+        public MapBeanNode(@Nullable String propertyName, Map<?, ?> map) {
+            super(propertyName, map);
+            this.map = map;
+        }
+
+        @Override
+        public boolean isIterable() {
+            return true;
+        }
+
+        @Override
+        public Iterator<BeanNode> getIterator() {
+            return Iterators.transform(map.entrySet().iterator(), new Function<Map.Entry<?, ?>, BeanNode>() {
+                @Override
+                public BeanNode apply(Map.Entry<?, ?> input) {
+                    return createChildNode(input.getKey().toString(), input.getValue());
```

Comment:
Should we have some check for `null` keys here to produce a better error message?

Reply:
Checking now.